includes/class-events-store.php

```diff
@@ -0,0 +1,592 @@
+		}
+
+		return $job;
+	}
+
+	/**
+	 * PLUGIN UTILITY METHODS
+	 */
+
+	/**
+	 * Retrieve jobs given a set of parameters
+	 *
+	 * @param array $args
+	 * @return array|false
+	 */
+	public function get_jobs( $args ) {
+		global $wpdb;
+
+		if ( ! isset( $args['quantity'] ) || ! is_numeric( $args['quantity'] ) ) {
+			$args['quantity'] = 100;
+		}
+
+		if ( isset( $args['page'] ) ) {
+			$page  = max( 0, $args['page'] - 1 );
+			$offset = $page * $args['quantity'];
+		} else {
+			$offset = 0;
+		}
+
+		// Do not sort, otherwise index isn't used
+		$jobs = $wpdb->get_results( $wpdb->prepare( "SELECT * FROM {$this->get_table_name()} WHERE status = %s LIMIT %d,%d;", $args['status'], $offset, $args['quantity'] ), 'OBJECT' );
+
+		if ( is_array( $jobs ) ) {
+			$jobs = array_map( array( $this, 'format_job' ), $jobs );
+		} else {
+			$jobs = false;
+		}
+
+		return $jobs;
+	}
+
+	/**
+	 * Retrieve a single event by ID, or by a combination of its timestamp, instance identifier, and either action or the action's hashed representation
+	 *
+	 * @param  array $attrs Array of event attributes to query by
+	 * @return object|false
+	 */
+	public function get_job( $attrs ) {
+		global $wpdb;
+
+		// Validate basic inputs
+		if ( ! is_array( $attrs ) || empty( $attrs ) ) {
+			return false;
+		}
+
+		// Validate requested status
+		$allowed_status = $this->get_allowed_statuses();
+		$allowed_status[] = 'any';
+
+		if ( ! isset( $attrs['status'] ) || ! in_array( $attrs['status'], $allowed_status, true ) ) {
+			$attrs['status'] = self::STATUS_PENDING;
+		}
+
+		// Validate attributes provided to query for a post
+		$query              = null;
+		$query_replacements = array();
+
+		if ( isset( $attrs['ID'] ) ) {
+			$query                = "SELECT * FROM {$this->get_table_name()} WHERE ID = %d";
+			$query_replacements[] = $attrs['ID'];
+		} else {
+			// Need a timestamp, an instance, and either an action or its hashed representation
+			if ( ! isset( $attrs['timestamp'] ) || ! isset( $attrs['instance'] ) ) {
+				return false;
+			} elseif ( ! isset( $attrs['action'] ) && ! isset( $attrs['action_hashed'] ) ) {
+				return false;
+			}
+
+			// Build query
+			if ( isset( $attrs['action'] ) ) {
+				$action_column = 'action';
+				$action_value  = $attrs['action'];
+			} else {
+				$action_column = 'action_hashed';
+				$action_value  = $attrs['action_hashed'];
+			}
+
+			// Do not sort, otherwise index isn't used
+			$query                = "SELECT * FROM {$this->get_table_name()} WHERE timestamp = %d AND {$action_column} = %s AND instance = %s";
+			$query_replacements[] = $attrs['timestamp'];
+			$query_replacements[] = $action_value;
+			$query_replacements[] = $attrs['instance'];
+		}
+
+		// Final query preparations
+		if ( 'any' !== $attrs['status'] ) {
+			$query                .= ' AND status = %s';
+			$query_replacements[]  = $attrs['status'];
+		}
+
+		$query .= ' LIMIT 1';
+
+		$query = $wpdb->prepare( $query, $query_replacements );
```

Comment:
The `query_replacements` approach seems a bit fragile. I think we should `prepare` in context (i.e. whenever we build each SQL statement).

If anything, `$query` should be an array of prepared SQL statements which is `implode`-ed with a space.